fedimint-client/src/sm/executor.rs

```diff
@@ -0,0 +1,614 @@
+                            },
+                        ),
+                    ]
+                }
+                MockStateMachine::ReceivedNonNull(prev_val) => {
+                    let prev_val = *prev_val;
+                    let mut receiver = context.broadcast.subscribe();
+                    vec![StateTransition::new(
+                        async move {
+                            loop {
+                                let val = receiver.recv().await.unwrap();
+                                if val == prev_val {
+                                    trace!("State transition ReceivedNonNull->Final");
+                                    break;
+                                }
+                            }
+                        },
+                        |_dbtx, (), _state| Box::pin(async move { MockStateMachine::Final }),
+                    )]
+                }
+                MockStateMachine::Final => {
+                    vec![]
+                }
+            }
+        }
+
+        fn operation_id(&self) -> OperationId {
+            [0u8; 32]
+        }
+    }
+
+    impl IntoDynInstance for MockStateMachine {
+        type DynType = DynState;
+
+        fn into_dyn(self, instance_id: ModuleInstanceId) -> Self::DynType {
+            DynState::from_typed(instance_id, self)
+        }
+    }
+
+    #[derive(Debug, Clone)]
+    struct MockContext {
+        broadcast: tokio::sync::broadcast::Sender<u64>,
+    }
+
+    impl IntoDynInstance for MockContext {
+        type DynType = DynContext;
+
+        fn into_dyn(self, instance_id: ModuleInstanceId) -> Self::DynType {
+            DynContext::from_typed(instance_id, self)
+        }
+    }
+
+    impl Context for MockContext {}
+
+    impl ClientModule for MockClientModule {
+        type StateMachineContext = MockContext;
+        type States = MockStateMachine;
+
+        fn decoder(&self) -> Decoder {
+            let mut decoder_builder = Decoder::builder();
+            decoder_builder.with_decodable_type::<MockStateMachine>();
+            decoder_builder.build()
+        }
+
+        fn context(&self) -> Self::StateMachineContext {
+            MockContext {
+                broadcast: self.0.clone(),
+            }
+        }
+    }
+
+    async fn get_executor(tg: &mut TaskGroup) -> (Executor, Sender<u64>, Database) {
+        let (broadcast, _) = tokio::sync::broadcast::channel(10);
+        let module = MockClientModule(broadcast.clone());
+
+        let decoders = ModuleDecoderRegistry::new(vec![(42, module.decoder())]);
+        let db = Database::new(MemDatabase::new(), decoders);
+
+        let mut executor_builder = Executor::builder();
+        executor_builder.with_module::<MockClientModule>(42, module);
+        let executor = executor_builder.build(tg, db.clone(), Arc::new(())).await;
+
+        info!("Initialized test executor");
+        (executor, broadcast, db)
+    }
+
+    #[tokio::test]
+    #[tracing_test::traced_test]
```

Comment:
What does this do? Don't see any assertions about log output ...

Reply:
I wanted to see log output, what did we use for tests elsewhere?

Reply:
I always just use `-- --nocapture`. This is good for now.